### ISSUE TITLE
redesign of Dropout

### ISSUE THREAD

**Issue by dvolgyes** [pytorch/pytorch#46184]
## 🚀 Feature
Well, the current Dropout* implementations are confusing, and at the same time miss feature (rescaling).
Unlike Cond*d, BatchNorm*d, etc., Dropout has three variants: Dropout, Dropout2d, Dropout3d.
The Dropout can drop any tensor element with p probability. On the other hand, the Dropout2d/3d
drops complete feature subtensors.
The misleading part is that the feature channel is always the second channel, regardless of dimensions.
So, if the input is 6 dimensional, or 3, they still drop by the second dimension, and there is no shape check at all.
Also, they have identical implementation.
So 2d/3d differentiation makes no sense, if anything, it could be called DropoutChannel or DropoutFeature.

There is also a missing feature, namely, rescaling the input with the dropout probability. 
As a matter of fact, i would prefer rescaling not with `p`, but actually with
```total elements / remaining elements```.

The reason is this: you can assume cases when sometimes 2 channels are kept, sometimes 3, and rescaling based
on the probability, not on the actual kept channels lead to 50% difference in output. Of course, with higher number of channels, the variance decreases, but if the dropout is redesigned anyway, this could be taken into account.
E.g. rescale = None | "prob" (default) | "kept" 
or something like this (instead of introducing a True/False parameter which is hard to change later).

Note that this rescaling is not the same as the AlphaDropout, and both of them are required.
For instance, tensorflow implements rescaling, and has a separate AlphaDropout too.

## Motivation

The current situation is confusing.

## Pitch

- clarify the documentation code
- add deprecation to Dropout2d / Dropout3d, and remove them later, or rename, like DropoutChannel
- introduce new class, preferably DropoutNd or similar, where you can specify any dimension,
  or you can specify a specific noise shape (e.g. rows or columns in an image, e.g. tensorflow's noise shape)
- unify defaults:  in the functional interface, Alpha dropout has a different default than the regular dropout
  for the ```training``` parameter (False and True, respectively).

## Alternatives

Not preferred, but clarified documentation with examples how to use Dropout 2d to be used for arbitrary dimensions, 
something like this:
```
# x.shape == (A1,A2,...,An,C, B1,B2, ...Bn)
x2=x.view(A1*A2*...,C, -1)
y=Dropout2d(x2).view(x.shape)
```

**Comment by dvolgyes** [comment]
Before selecting new name,  also consider the discussion in   #41243.
